### DISCUSSION TITLE
Upload historic NuGet packages without RepositoryUrl

### DISCUSSION THREAD

**Discussion by MaxHorstmann** [community/community#24920] [Actions]
<p>We would like to upload historic NuGet packages from our current package store to GitHub Packages. This would allow us to e.g. build hotfixes for older (but still supported) releases in GitHub Actions without having to first upgrade package dependencies and fix breaking changes.</p>
<p>Unfortunately, we can’t upload those old <code>.nupkg</code> files as-is because, <strong><a href="https://docs.microsoft.com/en-us/nuget/reference/nuspec#optional-metadata-elements" rel="noopener nofollow ugc">contrary to the nuspec spec</a></strong>, <code>Repository</code> is treated as a <strong>mandatory</strong> field in GitHub Packages.</p>
<pre><code class="lang-auto">$ dotnet nuget push MyOldPackage.0.1.17.nupkg -k &lt;token&gt; -s https://nuget.pkg.github.com/MyOrg/index.json
...
warn : No destination repository detected. Ensure the source project has a 'RepositoryUrl' property defined. If you're using a nuspec file, ensure that it has a repository element with the required 'type' and 'url' attributes.
</code></pre>
<p>How can we upload historic NuGet packages?</p>
<p>Two potential solutions come to mind: a change in GitHub Packages to no longer require <code>RepositoryUrl</code> (<a href="https://github.community/t/push-nuget-package-from-difference-owner/131732">supposedly planned</a>), or some mechanism to provide <code>RepositoryUrl</code> as a parameter to <code>nuget push</code> outside of the <code>.nupkg</code> file. Something like <code>nuget push --repository-url=...</code>.</p>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/maxhorstmann">@MaxHorstmann</a>,</p>
<aside class="quote no-group" data-username="MaxHorstmann" data-post="1" data-topic="166244">
<div class="title">
<div class="quote-controls"></div>
 MaxHorstmann:</div>
<blockquote>
<p>How can we upload historic NuGet packages?</p>
</blockquote>
</aside>
<p>I’ve written a .NET Core tool for working with the GitHub Package Registry, that can be used to upload historic NuGet packages.</p>
<p>You can install the tool like this:</p>
<pre><code class="lang-auto">$ dotnet tool install gpr -g
</code></pre>
<p>When pushing a package, you can provide a <code>RepositoryUrl</code> (like you suggested). You can use it like this:</p>
<pre><code class="lang-auto">$ gpr push path/to/*.nupkg --repository &lt;OWNER/REPO_NAME&gt; -k &lt;PAT&gt;
</code></pre>
<p>You can also do a batch upload like this:</p>
<pre><code class="lang-auto">$ gpr push 'path/to/**/*.nupkg' --repository &lt;OWNER/REPO_NAME&gt; -k &lt;PAT&gt;
</code></pre>
<p>(Your PAT will need the <code>repo</code> and <code>read:packages</code> scopes)</p>
<p>You can find more information about the tool here:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jcansdale/gpr" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:140/140;"><img src="https://avatars.githubusercontent.com/u/11719160?s=400&amp;amp;v=4" class="thumbnail" width="140" height="140"></div>

<h3><a href="https://github.com/jcansdale/gpr" target="_blank" rel="noopener nofollow ugc">jcansdale/gpr</a></h3>


  <p><span class="label1">A .NET Core tool for working with the GitHub Package Registry</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Please let me know if you have any questions!</p>

**Comment by MaxHorstmann** [comment]
<p>Outstanding- thank you very much! :clap:</p>
